### PR TITLE
THREESCALE-9570: Fix service plans scope on index page

### PR DESCRIPTION
**Pull request by mayorova**
**What this PR does / why we need it**:

Fix the issue where the index of service plans on a product include the service plans from other products.
Introduced by this change: https://github.com/3scale/porta/pull/3185/commits/99c6cb029f773831d287e4fdf1c9672c2f16d7b9#diff-2066514485e54295d6b177e8f030e3d8e547f73abbe2ae019981bac031e2232bR5

**Which issue(s) this PR fixes** 

https://issues.redhat.com/browse/THREESCALE-9570

**Verification steps** 

1. Create multiple products
2. Enable service plans in Audience > Usage Rules
3. Go to a product > Subscriptions > Service Plans
4. Make sure that only the service plans for the selected service are listed

**Special notes for your reviewer**:

### CODE REVIEW

**Review by jlledom — APPROVED** [review]

File: app/presenters/api/service_plans_presenter.rb [hunk 1]
```diff
@@ -1,8 +1,8 @@
 # frozen_string_literal: true
 
 class Api::ServicePlansPresenter < PlansBasePresenter
-  def initialize(service:, user:, params: {})
-    super(collection: service.provider.service_plans, user: user, params: params)
+  def initialize(service:, collection:, user:, params: {})
+    super(collection: collection, user: user, params: params)
```
Comment (josemigallas) [inline app/presenters/api/service_plans_presenter.rb:5]: I don't think we need to pass collection, just adjust the scope:
```suggestion
  def initialize(service:, user:, params: {})
    super(collection: service.service_plans, user: user, params: params)
```

As a side note, `#collection` does pretty much the same as `Api::PlansBaseController#find_service` so it could be cleaned up .

**Review by josemigallas — CHANGES_REQUESTED** [review]

**Comment by josemigallas** [comment]
Since service plans is now scoped, the following step needs changing:
```
When "an admin is in the service plans page" do
  visit admin_service_service_plans_path(@product || @service || default_service)
end
```
And I would use it in place of this "When I go to..." path: https://github.com/3scale/porta/blob/b38ca6c85353c0a01ef9689f4b188adb198abad3/features/support/paths.rb#L265-L266

File: app/presenters/api/service_plans_presenter.rb (hunk 1 above)
Comment (mayorova) [reply app/presenters/api/service_plans_presenter.rb:5]: I guess the idea here is that the presenter is not responsible for figuring out what elements of the collection to show, it's the controller's responsibility.

Like, if there was some restriction on what plans the current user can see, the filtering would be made by the controller, and not the presenter.

File: features/step_definitions/api/services/service_plans_steps.rb [hunk 2]
```diff
@@ -106,3 +106,18 @@
...
+end
+
+Then "only service plans of the current service are listed" do
+  FactoryBot.create(:service_plan, issuer: default_service, name: 'An extra plan')
+  FactoryBot.create(:service_plan, issuer: default_service, name: 'Another extra plan')
+
+  @plans = default_service.service_plans
+  visit admin_service_service_plans_path(default_service)
+  assert_plans_table @plans
+end
```
[the remark this answers is not in the document]
Comment (mayorova) [reply features/step_definitions/api/services/service_plans_steps.rb:123]: Yeah, OK, good.

The only thing is that I wanted to add some extra service plans, to make sure they are all visible, not only the default one, because if it's just "Default" - it could be coming from another service, as this is the default name of the service plan.

File: app/presenters/api/service_plans_presenter.rb (hunk 1 above)
Comment (mayorova) [reply app/presenters/api/service_plans_presenter.rb:5]: I tried to do
```
  def collection
    service.service_plans
  end
```

in `Api::ServicePlansController`, but there is an issue with the order of [these two hooks](https://github.com/3scale/porta/blob/scope-service-plans-by-product/app/controllers/api/plans_base_controller.rb#L10-L11).

In order for some controller methods to work like this, `find_service` and `find_plan` would need to be swapped. But some others are broken after this swapping, specifically updating the plan (as its URL does not include the `service_id`, so `service_id` is [derived from the service plan](https://github.com/3scale/porta/blob/scope-service-plans-by-product/app/controllers/api/plans_base_controller.rb#L56))

It would be easy if all URLs would be prefixed by service ID, but it's not the case.

So, in my opinion, this refactoring is not worth it, as it may break some existing functionality. Or well, we can do it in another PR, as it's not necessary for fixing this issue.

**Comment by mayorova** [comment]
@josemigallas Thanks for the comments!

> Since service plans is now scoped, the following step needs changing:

Well, not that it's "now scoped", it has always been that way.

I agree that it would be useful to have more flexibility in cucumbers, to allow visiting service plans index for other services, not only the default one. But it is not required for testing this bug fix.

How about I open another JIRA, that could include: 
- refactoring the `Api::ServicePlansController` and `Api::PlansBaseController` and their respective presenters
- cucumbers related to `service_plans.feature`

Basically, what you've suggested.

Cause I want to fix and merge this bug ASAP (as I think it's pretty bad), and we can refactor at a later stage.

**Review by jlledom — APPROVED** [review]

File: features/api/services/service_plans.feature [hunk 3]
```diff
@@ -15,6 +15,13 @@ Feature: Service plans index page
       Given a product
       Then an admin is able to see its service plans
 
+    Scenario: Service plans are scoped by service
+      Given a product
+      And there are other services with service plans
+      And the product has multiple service plans
+      When an admin is on the service plans page of the product
+      Then only service plans of the product are listed
+
```
Comment (jlledom) [inline features/api/services/service_plans.feature:24]: I think this is better tested in a unit test.

Comment (mayorova) [reply features/api/services/service_plans.feature:24]: An integration test you mean? Or unit test of the controller, however you want to call it.

Yes, that's what I started doing, but then I realized that the page initially has just a single `div`, the actual plans table is rendered via JavaScript

https://github.com/3scale/porta/blob/master/app/views/api/service_plans/index.html.slim

https://github.com/3scale/porta/blob/master/app/views/shared/plans/_plans_index_main_section.html.slim

At most we could test (in an integration test) that the JSON injected in the div's data is correct, but I think it's a big hacky. I preferred to do a full UI test for that.

Comment (jlledom) [reply features/api/services/service_plans.feature:24]: I meant a unit test for `Api::ServicePlansPresenter`. As you said, presenters are not the place to run queries, but since the query is there, it should be tested there, IMO.

Comment (mayorova) [reply features/api/services/service_plans.feature:24]: Ah, gotcha. Yes, makes sense.

Comment (mayorova) [reply features/api/services/service_plans.feature:24]: Added here: https://github.com/3scale/porta/pull/3479/commits/07da04365be28fc4b3a4747694dee8035e1d9b19

**Review by jlledom — APPROVED** [review]

**Review by josemigallas — APPROVED** [review]
